Use `ifNil:` in `isShorter:than:`.

fniephaus opened the pull request:
I'd argue this is slightly more idiomatic and it seems that this is also [supported in SOM](https://github.com/SOM-st/SOM/blob/f11313b14e6b85c1b5c5567e5b8cedc4b8e814fd/Smalltalk/Object.som#L51). In fact, the Squeak/Smalltalk compiler turns `ifNil:` into an identity check (`== nil`) plus jump rather than sending `isNil` to the receiver and then jump.

I noticed this some time ago when I looked into excessive splitting in TruffleSqueak ([here's 
a recent example](https://gist.github.com/TruffleSqueak-Bot/0fb5cbff18d9f47b3305a4d1879aa2bc#file-9ce8a19c-list-trace-log)). It turned out that the additional `isNil` send causes these splits.

Surely, there are ways to fix the splitting issue in TruffleSqueak. I just thought I open this anyway because, again, it is more idiomatic and because of that, also specifically optimized by the Squeak/Smalltalk compiler.

fniephaus committed:
Use `ifNil:` in `isShorter:than:`.

Instead of `isNil ifTrue:`.

smarr commented:
Looking at the other languages, I think the change should be fine.

Happy to merge it, but would prefer to keep the SOM and SOMns benchmarks in sync with the Smalltalk ones.

And yeah, the splitting issue, you probably still want to fix :)

fniephaus committed:
Use ifNil: in List.som

fniephaus commented:
Oh right, the SOM impl is separate. For some reason, I thought it’s derived from the Smalltalk code. I pushed another commit.

smarr committed:
Use ifNil: in List.ns

Signed-off-by: Stefan Marr <git@stefan-marr.de>

eregon commented:
There are a few more `isNil ifTrue:`, probably it would be good to change them too:
https://github.com/search?q=repo%3Asmarr%2Fare-we-fast-yet%20%2FisNil%20ifTrue%3A%2F&type=code

krono commented:
Note that some of these are newspeak, but the reasoning shouldn't be different.

krono commented:
The reason of `ifNil:` _not_ being used yet is that the original SOM, from which the benchmarks where pulled and built upon, did not support `ifNil:` back then.
Times are different now :)

fniephaus commented:
> There are a few more isNil ifTrue:, probably it would be good to change them too:
https://github.com/search?q=repo%3Asmarr%2Fare-we-fast-yet%20%2FisNil%20ifTrue%3A%2F&type=code

After looking at other usages, I'm not so sure anymore how far we want to go here. There are, for example, `isNil ifTrue: [] ifFalse: []` usages, which I would change to `ifNil: [] ifNotNil: []`, but that's not supported in SOM. Then, there also are usages of `notNil ifTrue: []`, which should probably be changed to `ifNotNil: []`, which is supported in SOM.

Any thoughts?

smarr commented:
Yeah, I pushed the SOMns commit, mostly to indicate that it's also separate code.
And, I didn't immediately merge it, because as @eregon pointed out, it would make things inconsistent. When looking at what we did for the other languages, it seemed desirable to apply this consistently to all benchmarks.

That some variations are currently missing from SOM seems only a small hurdle. Though, it's a bit of extra work. @OctaveLarose, for SOM-RS, did you implement special handling for `Boolean>>#ifNil:*`? I think I may have it for some of the SOMs already but not consistently.

OctaveLarose commented:
som-rs doesn't have `ifNil`, no. But it looks like it could be a speedup in a few benchmarks, and I could easily implement it. A lot of the code takes inspiration from PySOM's design, and I don't think it has any primitives for it, either.

It sounds like a good change, though! I'd not realized it was a common case.

smarr committed:
Apply change more globally

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Whitespace

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Add missing parentheses

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented:
The update for SOM is here: https://github.com/SOM-st/SOM/pull/123

smarr committed:
Use the dev branch of SOMns

The dev branch is really the main branch. The release branch is very old…
Also switch to Ubuntu 24.04 and set JAVA_HOME.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented:
Performance difference on the SOM implementations without specific support for `ifNil:*` variants: https://rebench.dev/Are-We-Fast-Yet/compare/2e1a50989356d8beb32c55393c8c2087d7f77707..a31e167bf1d6812a2251e10321d81abf0a3fab54

It looks a bit noisy, and there might be some extra overhead compared to the inlined `isNil ifTrue:*` version on List and Havlak.

With inlining of `ifNil:*` and variants, it seems a bit noisy and unclear, but no major change:
https://rebench.dev/Are-We-Fast-Yet/compare/2e1a50989356d8beb32c55393c8c2087d7f77707..807f8d81edef384d50638b82b80da6cb0f9e4775
